From b9011b35054982bf20760a655ec0504c223251c3 Mon Sep 17 00:00:00 2001
From: Jeevaka Prabu Badrappan <jeevaka.badrappan@intel.com>
Date: Fri, 15 Oct 2021 08:46:05 +0530
Subject: [PATCH] Revert "FROMLIST: KVM: mmu: introduce new gfn_to_pfn_page
 functions"

This reverts commit fbfe1e70219a16e921fdf22afb997eb72df16433.
Reverting as it causes device reboot during Android VM boot.

Signed-off-by: Jeevaka Prabu Badrappan <jeevaka.badrappan@intel.com>
---
 arch/x86/kvm/mmu/mmu.c   |   1 -
 include/linux/kvm_host.h |  17 ----
 virt/kvm/kvm_main.c      | 186 +++++++++++----------------------------
 3 files changed, 51 insertions(+), 153 deletions(-)

diff --git a/arch/x86/kvm/mmu/mmu.c b/arch/x86/kvm/mmu/mmu.c
index b6ec2a270dba..8fb78709c283 100644
--- a/arch/x86/kvm/mmu/mmu.c
+++ b/arch/x86/kvm/mmu/mmu.c
@@ -3710,7 +3710,6 @@ static bool try_async_pf(struct kvm_vcpu *vcpu, bool prefault, gfn_t gfn,
 	if (is_guest_mode(vcpu) && !kvm_is_visible_memslot(slot)) {
 		*pfn = KVM_PFN_NOSLOT;
 		*writable = false;
-		*page = NULL;
 		return false;
 	}
 
diff --git a/include/linux/kvm_host.h b/include/linux/kvm_host.h
index 9a41375cf8d4..947710eea946 100644
--- a/include/linux/kvm_host.h
+++ b/include/linux/kvm_host.h
@@ -768,19 +768,6 @@ kvm_pfn_t __gfn_to_pfn_memslot(struct kvm_memory_slot *slot, gfn_t gfn,
 			       bool atomic, bool *async, bool write_fault,
 			       bool *writable, hva_t *hva);
 
-kvm_pfn_t gfn_to_pfn_page(struct kvm *kvm, gfn_t gfn, struct page **page);
-kvm_pfn_t gfn_to_pfn_page_prot(struct kvm *kvm, gfn_t gfn,
-			       bool write_fault, bool *writable,
-			       struct page **page);
-kvm_pfn_t gfn_to_pfn_page_memslot(struct kvm_memory_slot *slot,
-				  gfn_t gfn, struct page **page);
-kvm_pfn_t gfn_to_pfn_page_memslot_atomic(struct kvm_memory_slot *slot,
-					 gfn_t gfn, struct page **page);
-kvm_pfn_t __gfn_to_pfn_page_memslot(struct kvm_memory_slot *slot,
-				    gfn_t gfn, bool atomic, bool *async,
-				    bool write_fault, bool *writable,
-				    hva_t *hva, struct page **page);
-
 void kvm_release_pfn_clean(kvm_pfn_t pfn);
 void kvm_release_pfn_dirty(kvm_pfn_t pfn);
 void kvm_set_pfn_dirty(kvm_pfn_t pfn);
@@ -863,10 +850,6 @@ struct kvm_memslots *kvm_vcpu_memslots(struct kvm_vcpu *vcpu);
 struct kvm_memory_slot *kvm_vcpu_gfn_to_memslot(struct kvm_vcpu *vcpu, gfn_t gfn);
 kvm_pfn_t kvm_vcpu_gfn_to_pfn_atomic(struct kvm_vcpu *vcpu, gfn_t gfn);
 kvm_pfn_t kvm_vcpu_gfn_to_pfn(struct kvm_vcpu *vcpu, gfn_t gfn);
-kvm_pfn_t kvm_vcpu_gfn_to_pfn_page_atomic(struct kvm_vcpu *vcpu, gfn_t gfn,
-					  struct page **page);
-kvm_pfn_t kvm_vcpu_gfn_to_pfn_page(struct kvm_vcpu *vcpu, gfn_t gfn,
-				   struct page **page);
 int kvm_vcpu_map(struct kvm_vcpu *vcpu, gpa_t gpa, struct kvm_host_map *map);
 int kvm_map_gfn(struct kvm_vcpu *vcpu, gfn_t gfn, struct kvm_host_map *map,
 		struct gfn_to_pfn_cache *cache, bool atomic);
diff --git a/virt/kvm/kvm_main.c b/virt/kvm/kvm_main.c
index 08f93902abf5..999a92e3d421 100644
--- a/virt/kvm/kvm_main.c
+++ b/virt/kvm/kvm_main.c
@@ -2046,9 +2046,9 @@ static inline int check_user_page_hwpoison(unsigned long addr)
  * only part that runs if we can in atomic context.
  */
 static bool hva_to_pfn_fast(unsigned long addr, bool write_fault,
-			    bool *writable, kvm_pfn_t *pfn,
-			    struct page **page)
+			    bool *writable, kvm_pfn_t *pfn)
 {
+	struct page *page[1];
 
 	/*
 	 * Fast pin a writable pfn only if it is a write fault request
@@ -2059,7 +2059,7 @@ static bool hva_to_pfn_fast(unsigned long addr, bool write_fault,
 		return false;
 
 	if (get_user_page_fast_only(addr, FOLL_WRITE, page)) {
-		*pfn = page_to_pfn(*page);
+		*pfn = page_to_pfn(page[0]);
 
 		if (writable)
 			*writable = true;
@@ -2074,9 +2074,10 @@ static bool hva_to_pfn_fast(unsigned long addr, bool write_fault,
  * 1 indicates success, -errno is returned if error is detected.
  */
 static int hva_to_pfn_slow(unsigned long addr, bool *async, bool write_fault,
-			   bool *writable, kvm_pfn_t *pfn, struct page **page)
+			   bool *writable, kvm_pfn_t *pfn)
 {
 	unsigned int flags = FOLL_HWPOISON;
+	struct page *page;
 	int npages = 0;
 
 	might_sleep();
@@ -2089,7 +2090,7 @@ static int hva_to_pfn_slow(unsigned long addr, bool *async, bool write_fault,
 	if (async)
 		flags |= FOLL_NOWAIT;
 
-	npages = get_user_pages_unlocked(addr, 1, page, flags);
+	npages = get_user_pages_unlocked(addr, 1, &page, flags);
 	if (npages != 1)
 		return npages;
 
@@ -2099,11 +2100,11 @@ static int hva_to_pfn_slow(unsigned long addr, bool *async, bool write_fault,
 
 		if (get_user_page_fast_only(addr, FOLL_WRITE, &wpage)) {
 			*writable = true;
-			put_page(*page);
-			*page = wpage;
+			put_page(page);
+			page = wpage;
 		}
 	}
-	*pfn = page_to_pfn(*page);
+	*pfn = page_to_pfn(page);
 	return npages;
 }
 
@@ -2118,6 +2119,13 @@ static bool vma_is_valid(struct vm_area_struct *vma, bool write_fault)
 	return true;
 }
 
+static int kvm_try_get_pfn(kvm_pfn_t pfn)
+{
+	if (kvm_is_reserved_pfn(pfn))
+		return 1;
+	return get_page_unless_zero(pfn_to_page(pfn));
+}
+
 static int hva_to_pfn_remapped(struct vm_area_struct *vma,
 			       unsigned long addr, bool *async,
 			       bool write_fault, bool *writable,
@@ -2157,6 +2165,26 @@ static int hva_to_pfn_remapped(struct vm_area_struct *vma,
 		*writable = pte_write(*ptep);
 	pfn = pte_pfn(*ptep);
 
+	/*
+	 * Get a reference here because callers of *hva_to_pfn* and
+	 * *gfn_to_pfn* ultimately call kvm_release_pfn_clean on the
+	 * returned pfn.  This is only needed if the VMA has VM_MIXEDMAP
+	 * set, but the kvm_get_pfn/kvm_release_pfn_clean pair will
+	 * simply do nothing for reserved pfns.
+	 *
+	 * Whoever called remap_pfn_range is also going to call e.g.
+	 * unmap_mapping_range before the underlying pages are freed,
+	 * causing a call to our MMU notifier.
+	 *
+	 * Certain IO or PFNMAP mappings can be backed with valid
+	 * struct pages, but be allocated without refcounting e.g.,
+	 * tail pages of non-compound higher order allocations, which
+	 * would then underflow the refcount when the caller does the
+	 * required put_page. Don't allow those pages here.
+	 */ 
+	if (!kvm_try_get_pfn(pfn))
+		r = -EFAULT;
+
 out:
 	pte_unmap_unlock(ptep, ptl);
 	*p_pfn = pfn;
@@ -2178,9 +2206,8 @@ static int hva_to_pfn_remapped(struct vm_area_struct *vma,
  * 2): @write_fault = false && @writable, @writable will tell the caller
  *     whether the mapping is writable.
  */
-static kvm_pfn_t hva_to_pfn(unsigned long addr, bool atomic,
-			    bool *async, bool write_fault, bool *writable,
-			    struct page **page)
+static kvm_pfn_t hva_to_pfn(unsigned long addr, bool atomic, bool *async,
+			bool write_fault, bool *writable)
 {
 	struct vm_area_struct *vma;
 	kvm_pfn_t pfn = 0;
@@ -2189,14 +2216,13 @@ static kvm_pfn_t hva_to_pfn(unsigned long addr, bool atomic,
 	/* we can do it either atomically or asynchronously, not both */
 	BUG_ON(atomic && async);
 
-	if (hva_to_pfn_fast(addr, write_fault, writable, &pfn, page))
+	if (hva_to_pfn_fast(addr, write_fault, writable, &pfn))
 		return pfn;
 
 	if (atomic)
 		return KVM_PFN_ERR_FAULT;
 
-	npages = hva_to_pfn_slow(addr, async, write_fault, writable,
-				 &pfn, page);
+	npages = hva_to_pfn_slow(addr, async, write_fault, writable, &pfn);
 	if (npages == 1)
 		return pfn;
 
@@ -2228,14 +2254,12 @@ static kvm_pfn_t hva_to_pfn(unsigned long addr, bool atomic,
 	return pfn;
 }
 
-kvm_pfn_t __gfn_to_pfn_page_memslot(struct kvm_memory_slot *slot,
-				    gfn_t gfn, bool atomic, bool *async,
-				    bool write_fault, bool *writable,
-				    hva_t *hva, struct page **page)
+kvm_pfn_t __gfn_to_pfn_memslot(struct kvm_memory_slot *slot, gfn_t gfn,
+			       bool atomic, bool *async, bool write_fault,
+			       bool *writable, hva_t *hva)
 {
 	unsigned long addr = __gfn_to_hva_many(slot, gfn, NULL, write_fault);
 
-	*page = NULL;
 	if (hva)
 		*hva = addr;
 
@@ -2258,153 +2282,45 @@ kvm_pfn_t __gfn_to_pfn_page_memslot(struct kvm_memory_slot *slot,
 	}
 
 	return hva_to_pfn(addr, atomic, async, write_fault,
-			  writable, page);
-}
-EXPORT_SYMBOL_GPL(__gfn_to_pfn_page_memslot);
-
-kvm_pfn_t gfn_to_pfn_page_prot(struct kvm *kvm, gfn_t gfn, bool write_fault,
-			       bool *writable, struct page **page)
-{
-	return __gfn_to_pfn_page_memslot(gfn_to_memslot(kvm, gfn), gfn, false,
-					 NULL, write_fault, writable, NULL,
-					 page);
-}
-EXPORT_SYMBOL_GPL(gfn_to_pfn_page_prot);
-
-kvm_pfn_t gfn_to_pfn_page_memslot(struct kvm_memory_slot *slot, gfn_t gfn,
-				  struct page **page)
-{
-	return __gfn_to_pfn_page_memslot(slot, gfn, false, NULL, true,
-					 NULL, NULL, page);
-}
-EXPORT_SYMBOL_GPL(gfn_to_pfn_page_memslot);
-
-kvm_pfn_t gfn_to_pfn_page_memslot_atomic(struct kvm_memory_slot *slot,
-					 gfn_t gfn, struct page **page)
-{
-	return __gfn_to_pfn_page_memslot(slot, gfn, true, NULL, true, NULL,
-					 NULL, page);
-}
-EXPORT_SYMBOL_GPL(gfn_to_pfn_page_memslot_atomic);
-
-kvm_pfn_t kvm_vcpu_gfn_to_pfn_page_atomic(struct kvm_vcpu *vcpu, gfn_t gfn,
-					  struct page **page)
-{
-	return gfn_to_pfn_page_memslot_atomic(
-			kvm_vcpu_gfn_to_memslot(vcpu, gfn), gfn, page);
-}
-EXPORT_SYMBOL_GPL(kvm_vcpu_gfn_to_pfn_page_atomic);
-
-kvm_pfn_t gfn_to_pfn_page(struct kvm *kvm, gfn_t gfn, struct page **page)
-{
-	return gfn_to_pfn_page_memslot(gfn_to_memslot(kvm, gfn), gfn, page);
-}
-EXPORT_SYMBOL_GPL(gfn_to_pfn_page);
-
-kvm_pfn_t kvm_vcpu_gfn_to_pfn_page(struct kvm_vcpu *vcpu, gfn_t gfn,
-				   struct page **page)
-{
-	return gfn_to_pfn_page_memslot(kvm_vcpu_gfn_to_memslot(vcpu, gfn),
-				       gfn, page);
-}
-EXPORT_SYMBOL_GPL(kvm_vcpu_gfn_to_pfn_page);
-
-static kvm_pfn_t ensure_pfn_ref(struct page *page, kvm_pfn_t pfn)
-{
-	if (page || is_error_pfn(pfn) || kvm_is_reserved_pfn(pfn))
-		return pfn;
-
-	/*
-	 * Certain IO or PFNMAP mappings can be backed with valid
-	 * struct pages, but be allocated without refcounting e.g.,
-	 * tail pages of non-compound higher order allocations, which
-	 * would then underflow the refcount when the caller does the
-	 * required put_page. Don't allow those pages here.
-	 */
-	if (get_page_unless_zero(pfn_to_page(pfn)))
-		return pfn;
-
-	return KVM_PFN_ERR_FAULT;
-}
-
-kvm_pfn_t __gfn_to_pfn_memslot(struct kvm_memory_slot *slot, gfn_t gfn,
-			       bool atomic, bool *async, bool write_fault,
-			       bool *writable, hva_t *hva)
-{
-	struct page *page;
-	kvm_pfn_t pfn;
-
-	pfn = __gfn_to_pfn_page_memslot(slot, gfn, atomic, async,
-					write_fault, writable, hva, &page);
-
-	return ensure_pfn_ref(page, pfn);
+			  writable);
 }
 EXPORT_SYMBOL_GPL(__gfn_to_pfn_memslot);
 
 kvm_pfn_t gfn_to_pfn_prot(struct kvm *kvm, gfn_t gfn, bool write_fault,
 		      bool *writable)
 {
-	struct page *page;
-	kvm_pfn_t pfn;
-
-	pfn = gfn_to_pfn_page_prot(kvm, gfn, write_fault, writable, &page);
-
-	return ensure_pfn_ref(page, pfn);
+	return __gfn_to_pfn_memslot(gfn_to_memslot(kvm, gfn), gfn, false, NULL,
+				    write_fault, writable, NULL);
 }
 EXPORT_SYMBOL_GPL(gfn_to_pfn_prot);
 
 kvm_pfn_t gfn_to_pfn_memslot(struct kvm_memory_slot *slot, gfn_t gfn)
 {
-	struct page *page;
-	kvm_pfn_t pfn;
-
-	pfn = gfn_to_pfn_page_memslot(slot, gfn, &page);
-
-	return ensure_pfn_ref(page, pfn);
+	return __gfn_to_pfn_memslot(slot, gfn, false, NULL, true, NULL, NULL);
 }
 EXPORT_SYMBOL_GPL(gfn_to_pfn_memslot);
 
 kvm_pfn_t gfn_to_pfn_memslot_atomic(struct kvm_memory_slot *slot, gfn_t gfn)
 {
-	struct page *page;
-	kvm_pfn_t pfn;
-
-	pfn = gfn_to_pfn_page_memslot_atomic(slot, gfn, &page);
-
-	return ensure_pfn_ref(page, pfn);
+	return __gfn_to_pfn_memslot(slot, gfn, true, NULL, true, NULL, NULL);
 }
 EXPORT_SYMBOL_GPL(gfn_to_pfn_memslot_atomic);
 
 kvm_pfn_t kvm_vcpu_gfn_to_pfn_atomic(struct kvm_vcpu *vcpu, gfn_t gfn)
 {
-	struct page *page;
-	kvm_pfn_t pfn;
-
-	pfn = kvm_vcpu_gfn_to_pfn_page_atomic(vcpu, gfn, &page);
-
-	return ensure_pfn_ref(page, pfn);
+	return gfn_to_pfn_memslot_atomic(kvm_vcpu_gfn_to_memslot(vcpu, gfn), gfn);
 }
 EXPORT_SYMBOL_GPL(kvm_vcpu_gfn_to_pfn_atomic);
 
 kvm_pfn_t gfn_to_pfn(struct kvm *kvm, gfn_t gfn)
 {
-	struct page *page;
-	kvm_pfn_t pfn;
-
-	pfn = gfn_to_pfn_page(kvm, gfn, &page);
-
-	return ensure_pfn_ref(page, pfn);
+	return gfn_to_pfn_memslot(gfn_to_memslot(kvm, gfn), gfn);
 }
 EXPORT_SYMBOL_GPL(gfn_to_pfn);
 
 kvm_pfn_t kvm_vcpu_gfn_to_pfn(struct kvm_vcpu *vcpu, gfn_t gfn)
 {
-	struct page *page;
-	kvm_pfn_t pfn;
-
-	pfn = kvm_vcpu_gfn_to_pfn_page(vcpu, gfn, &page);
-
-	return ensure_pfn_ref(page, pfn);
+	return gfn_to_pfn_memslot(kvm_vcpu_gfn_to_memslot(vcpu, gfn), gfn);
 }
 EXPORT_SYMBOL_GPL(kvm_vcpu_gfn_to_pfn);
 
-- 
2.40.1

